Shoko.Server/API/v3/Controllers/PlexController.cs

```diff
@@ -0,0 +1,203 @@
+    /// <returns></returns>
+    [HttpGet("Auth/Authenticate")]
+    public ActionResult<string> GetOAuthRequestUrl([FromQuery] bool force = false)
+    {
+        var helper = GetHelperForUser();
+
+        if (helper.IsAuthenticated && !force)
+            return BadRequest(AlreadyAuthenticated);
+
+        return helper.LoginUrl;
+    }
+
+    /// <summary>
+    /// Check if the current user is currently authenticated against the plex api.
+    /// </summary>
+    /// <returns></returns>
+    [HttpGet("Auth/IsAuthenticated")]
+    public ActionResult<bool> IsAuthenticated()
+    {
+        var helper = GetHelperForUser();
+
+        return helper.IsAuthenticated;
+    }
+
+    /// <summary>
+    /// Invalidate and remove the current plex authentication token.
+    /// </summary>
+    /// <returns></returns>
+    [HttpPost("Auth/InvalidateToken")]
+    public ActionResult<bool> InvalidateToken()
+    {
+        var helper = GetHelperForUser();
+
+        if (!helper.IsAuthenticated)
+            return false;
+
+        helper.InvalidateToken();
+
+        return true;
+    }
+
+    /// <summary>
+    /// Show all available server for the current user.
+    /// </summary>
+    /// <param name="userId"></param>
+    /// <returns></returns>
+    [HttpGet("Linking/AvailableServers")]
+    public ActionResult<List<PlexServer>> AvailableServers(int userId)
+    {
+        var helper = GetHelperForUser();
+
+        if (!helper.IsAuthenticated)
+            return BadRequest(NotAuthenticated);
+
+        var currentServerID = helper.ServerCache?.ClientIdentifier ?? "";
+        return helper.GetPlexServers()
+            .Select(device => new PlexServer(device, device.ClientIdentifier == currentServerID))
+            .ToList();
+    }
+
+    /// <summary>
+    /// Get the active server.
+    /// </summary>
+    /// <returns></returns>
+    [HttpGet("Linking/Server")]
+    public ActionResult<PlexServer> CurrentDevice()
+    {
+        var helper = GetHelperForUser();
+
+        if (!helper.IsAuthenticated)
+            return BadRequest(NotAuthenticated);
+
+        var currentServer = helper.ServerCache;
+        if (currentServer == null)
+            return BadRequest(NoSelectedServer);
+
+        return new PlexServer(currentServer, true);
+    }
+
+    /// <summary>
+    /// Select the active server.
+    /// </summary>
+    /// <param name="server"></param>
+    /// <returns></returns>
+    [HttpPut("Linking/Server")]
+    public ActionResult<PlexServer> UseDevice(PlexServer server)
+    {
+        var helper = GetHelperForUser();
+
+        if (!helper.IsAuthenticated)
+            return BadRequest(NotAuthenticated);
+
+        var previousServer = helper.ServerCache;
+        var currentServer = helper.GetPlexServers()
+            .FirstOrDefault(ser => ser.ClientIdentifier == server.ID);
+
+        if (currentServer == null || !currentServer.Provides.Split(',').Contains("server"))
+            return BadRequest(InvalidServerSelection);
```

Comment:
This is too much of the "is it a server" logic in the controller - can we move this out?

Reply:
It was doing the same in the `helper.UseServer` with the exception of it **silently failing** and not informing the user that the server **were not selected** because it was an invalid device to begin with.

If we can either change the helper to return the currently selected server (so we can check if it was updated) or throw inside the helper instead of silently returning then I'm fine with removing the check here, but I still want it to complain if the selected device is invalid as a form of user input validation and also just to inform the user that they tried to select an invalid device.

Reply:
This is a matter for consistency sake, and the problem if we want to look. At exposing the logic elsewhere we have to duplicate the logic, IMO we shouldn’t really have much decision making in controllers but defer that decision making to central classes that are agnostic of the request method.